### PR TITLE
wasm: add support for GOOS=wasip1

### PR DESCRIPTION
**Pull request by aykevl**
This is WIP, a preview of what we might want to add (see https://github.com/tinygo-org/tinygo/issues/3685).

This is bigger than you might expect, for two reasons:

1. This uses GOOS=wasip1 support in the Go standard library (added in Go 1.21) instead of using our old support for WASM that basically just pretended to be linux/arm as a workaround for lack of native support.
2. Support is entirely built into the compiler (instead of in a target JSON file), similar to other operating systems (linux, darwin, windows) but unlike our existing js/wasm and wasi/wasm support. I'm not entirely sure this is a good idea, although it does make sense if we expose this feature only using environment variables and not using the `-target` flag.

It's not fully working, I'll need to continue working on it later. But a basic "hello world" already compiles.

### CODE REVIEW

**Comment by deadprogram** [comment]
> This is bigger than you might expect

Perhaps we should move this to milestone for `0.30` release so we do not hold up a Go 1.21 compatible release in the meantime.

@aykevl @codefromthecrypt anyone else what do you think?

**Comment by aykevl** [comment]
Actually I think I can finish this PR quickly. It seems to pass all tests locally at least.

**Comment by aykevl** [comment]
Eventually I'd like to remove our old WASI support entirely (currently usable via `-target=wasi`), perhaps aliasing `-target=wasi` to the equivalent of `GOOS=wasip1 GOARCH=wasm` to not break backwards compatibility too badly. But for now it seems reasonable to keep support for both. It's not too difficult to support both at the same time: it's mainly just slightly more complicated `//go:build` lines and somewhat slower CI.

**Comment by deadprogram** [comment]
```
/root/project/src/runtime/os_wasip1.go:8:7: GOOS redeclared in this block
/root/project/src/runtime/os_linux.go:10:7: 	other declaration of GOOS
# runtime
/root/project/src/runtime/os_wasip1.go:8:7: GOOS redeclared in this block
/root/project/src/runtime/os_linux.go:10:7: 	other declaration of GOOS
make: *** [Makefile:449: smoketest] Error 1

Exited with code exit status 2
```

Actual fail in CircleCI build https://app.circleci.com/pipelines/github/tinygo-org/tinygo/9892/workflows/c400192b-ca42-4e4a-a87c-f7743e786c75/jobs/41137?invite=true#step-118-9

**Comment by aykevl** [comment]
Thanks, fixed now.

**Comment by aykevl** [comment]
Fixed an issue on `tinygo test os` (due to the renaming of stat_linux.go).

**Review by deadprogram — APPROVED** [review]
Looking good from my point of view.

@dgryski @codefromthecrypt @dkegel-fastly what do you think?

**Comment by deadprogram** [comment]
Last call for comments before merging this PR!

cc @dgryski @codefromthecrypt @dkegel-fastly @achille-roussel and anyone else who is interested.

**Review by achille-roussel — APPROVED** [review]
This looks great 🙌 it's actually a much smaller change than I would have expected, which is great 👍 

It might be useful to document when should users select `-target=wasi` vs `GOOS=wasip1` in a README or similar, I believe that the historical implementation still had a few features that were not supported, for example:
- https://github.com/tinygo-org/tinygo/pull/3699
- https://github.com/tinygo-org/tinygo/pull/3700

I imagine users will probably stubble upon those differences and be confused about why certain functionalities work in one WASI but not the other. I expect we would get questions like "when should we use -target=wasi vs GOOS=wasip1?" and recording those answers will likely reduce the maintenance load.

Thanks for doing this work!

**Comment by dkegel-fastly** [comment]
running my app's tests now...

**Comment by dkegel-fastly** [comment]
In preparation for testing this branch, I did a hopefully clean build of the latest dev branch, but it fails "make test", so maybe hold off until that's settled...? see slack.

**Comment by deadprogram** [comment]
@dkegel-fastly see https://github.com/tinygo-org/tinygo/issues/3871 to track that issue. Since it is not related to this PR, I am going to merge this one to keep things moving along.

**Comment by dgryski** [comment]
BTW, this seems to pass the test corpus just fine.

**Pull request merged**
